### PR TITLE
Feature pipsolar anh

### PR DESCRIPTION
**Pull request by andreashergert1984**
# What does this implement/fix? 

Quick description 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1084
  
# Test Environment

- [X] ESP32
- [X] ESP8266
- [ ] Windows
- [X] Mac OS
- [X] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

pipsolar:
    id: inverter0

sensor: 
  - platform: pipsolar
    pipsolar_id: inverter0
#QPIRI
    grid_rating_voltage:
      id: inverter0_grid_rating_voltage
      name: inverter0_grid_rating_voltage
    grid_rating_current:
      id: inverter0_grid_rating_current
      name: inverter0_grid_rating_current
    ac_output_rating_voltage:
      id: inverter0_ac_output_rating_voltage
      name: inverter0_ac_output_rating_voltage
    ac_output_rating_frequency:
      id: inverter0_ac_output_rating_frequency
      name: inverter0_ac_output_rating_frequency
    ac_output_rating_current:
      id: inverter0_ac_output_rating_current
      name: inverter0_ac_output_rating_current
    ac_output_rating_apparent_power:
      id: inverter0_ac_output_rating_apparent_power
      name: inverter0_ac_output_rating_apparent_power
    ac_output_rating_active_power:
      id: inverter0_ac_output_rating_active_power
      name: inverter0_ac_output_rating_active_power
    battery_rating_voltage:
      id: inverter0_battery_rating_voltage
      name: inverter0_battery_rating_voltage
    battery_recharge_voltage:
      id: inverter0_battery_recharge_voltage
      name: inverter0_battery_recharge_voltage
    battery_under_voltage:
      id: inverter0_battery_under_voltage
      name: inverter0_battery_under_voltage
    battery_bulk_voltage:
      id: inverter0_battery_bulk_voltage
      name: inverter0_battery_bulk_voltage
    battery_float_voltage:
      id: inverter0_battery_float_voltage
      name: inverter0_battery_float_voltage
    battery_type:
      id: inverter0_battery_type
      name: inverter0_battery_type
    current_max_ac_charging_current:
      id: inverter0_current_max_ac_charging_current
      name: inverter0_current_max_ac_charging_current
    current_max_charging_current:
      id: inverter0_current_max_charging_current
      name: inverter0_current_max_charging_current
    input_voltage_range:
      id: inverter0_input_voltage_range
      name: inverter0_input_voltage_range
    output_source_priority:
      id: inverter0_output_source_priority
      name: inverter0_output_source_priority
    charger_source_priority:
      id: inverter0_charger_source_priority
      name: inverter0_charger_source_priority
    parallel_max_num:
      id: inverter0_parallel_max_num
      name: inverter0_parallel_max_num
    machine_type:
      id: inverter0_machine_type
      name: inverter0_machine_type
    topology:
      id: inverter0_topology
      name: inverter0_topology
    output_mode:
      id: inverter0_output_mode
      name: inverter0_output_mode
    battery_redischarge_voltage:
      id: inverter0_battery_redischarge_voltage
      name: inverter0_battery_redischarge_voltage
    pv_ok_condition_for_parallel:
      id: inverter0_pv_ok_condition_for_parallel
      name: inverter0_pv_ok_condition_for_parallel
    pv_power_balance:
      id: inverter0_pv_power_balance
      name: inverter0_pv_power_balance
#QPIGS
    grid_voltage:
      id: inverter0_grid_voltage
      name: inverter0_grid_voltage
    grid_frequency:
      id: inverter0_grid_frequency
      name: inverter0_grid_frequency
    ac_output_voltage:
      id: inverter0_ac_output_voltage
      name: inverter0_ac_output_voltage
    ac_output_frequency:
      id: inverter0_ac_output_frequency
      name: inverter0_ac_output_frequency
    ac_output_apparent_power:
      id: inverter0_ac_output_apparent_power
      name: inverter0_ac_output_apparent_power
    ac_output_active_power:
      id: inverter0_ac_output_active_power
      name: inverter0_ac_output_active_power
    output_load_percent:
      id: inverter0_output_load_percent
      name: inverter0_output_load_percent
    bus_voltage:
      id: inverter0_bus_voltage
      name: inverter0_bus_voltage
    battery_voltage:
      id: inverter0_battery_voltage
      name: inverter0_battery_voltage
    battery_charging_current:
      id: inverter0_battery_charging_current
      name: inverter0_battery_charging_current
    battery_capacity_percent:
      id: inverter0_battery_capacity_percent
      name: inverter0_battery_capacity_percent
    inverter_heat_sink_temperature:
      id: inverter0_inverter_heat_sink_temperature
      name: inverter0_inverter_heat_sink_temperature
    pv_input_current_for_battery:
      id: inverter0_pv_input_current_for_battery
      name: inverter0_pv_input_current_for_battery
    pv_input_voltage:
      id: inverter0_pv_input_voltage
      name: inverter0_pv_input_voltage
    battery_voltage_scc:
      id: inverter0_battery_voltage_scc
      name: inverter0_battery_voltage_scc
    battery_discharge_current:
      id: inverter0_battery_discharge_current
      name: inverter0_battery_discharge_current
    battery_voltage_offset_for_fans_on:
      id: inverter0_battery_voltage_offset_for_fans_on
      name: inverter0_battery_voltage_offset_for_fans_on
    eeprom_version:
      id: inverter0_eeprom_version
      name: inverter0_eeprom_version
    pv_charging_power:
      id: inverter0_pv_charging_power
      name: inverter0_pv_charging_power

text_sensor:
  - platform: pipsolar
    pipsolar_id: inverter0
    device_mode:
      id: inverter0_device_mode
      name: inverter0_device_mode
    last_qpigs:
      id: inverter0_last_qpigs
      name: inverter0_last_qpigs
    last_qpiri:
      id: inverter0_last_qpiri
      name: inverter0_last_qpiri
    last_qmod:
      id: inverter0_last_qmod
      name: inverter0_last_qmod
    last_qflag:
      id: inverter0_last_qflag
      name: inverter0_last_qflag
binary_sensor:
  - platform: pipsolar
    pipsolar_id: inverter0
    add_sbu_priority_version:
      id: inverter0_add_sbu_priority_version
      name: inverter0_add_sbu_priority_version
    configuration_status:
      id: inverter0_configuration_status
      name: inverter0_configuration_status
    scc_firmware_version:
      id: inverter0_scc_firmware_version
      name: inverter0_scc_firmware_version
    load_status:
      id: inverter0_load_status
      name: inverter0_load_status
    battery_voltage_to_steady_while_charging:
      id: inverter0_battery_voltage_to_steady_while_charging
      name: inverter0_battery_voltage_to_steady_while_charging
    charging_status:
      id: inverter0_charging_status
      name: inverter0_charging_status
    scc_charging_status:
      id: inverter0_scc_charging_status
      name: inverter0_scc_charging_status
    ac_charging_status:
      id: inverter0_ac_charging_status
      name: inverter0_ac_charging_status
    charging_to_floating_mode:
      id: inverter0_charging_to_floating_mode
      name: inverter0_charging_to_floating_mode
    switch_on:
      id: inverter0_switch_on
      name: inverter0_switch_on
    dustproof_installed:
      id: inverter0_dustproof_installed
      name: inverter0_dustproof_installed
    silence_buzzer_open_buzzer:
      id: inverter0_silence_buzzer_open_buzzer
      name: inverter0_silence_buzzer_open_buzzer
    overload_bypass_function:
      id: inverter0_overload_bypass_function
      name: inverter0_overload_bypass_function
    lcd_escape_to_default:
      id: inverter0_lcd_escape_to_default
      name: inverter0_lcd_escape_to_default
    overload_restart_function:
      id: inverter0_overload_restart_function
      name: inverter0_overload_restart_function
    over_temperature_restart_function:
      id: inverter0_over_temperature_restart_function
      name: inverter0_over_temperature_restart_function
    backlight_on:
      id: inverter0_backlight_on
      name: inverter0_backlight_on
switch:
  - platform: pipsolar
    pipsolar_id: inverter0
    output_source_priority_utility:
      name: inverter0_output_source_priority_utility
    output_source_priority_solar:
      name: inverter0_output_source_priority_solar
    output_source_priority_battery:
      name: inverter0_output_source_priority_battery
    input_voltage_range:
      name: inverter0_input_voltage_range
    pv_ok_condition_for_parallel:
      name: inverter0_pv_ok_condition_for_parallel
    pv_power_balance:
      name: inverter0_pv_power_balance

output:
  - platform: pipsolar
    pipsolar_id: inverter0
    battery_recharge_voltage:
      id: inverter0_battery_recharge_voltage_out
```

# Explain your changes

Describe your changes here to communicate to the maintainers **why we should accept this pull request**.
Very important to fill if no issue linked

This components adds the possibility to read and interact to pip solar compatible pv inverters.
A huge amount of sensors and some actors are possible.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/pipsolar/pipsolar.cpp [hunk 1]
```diff
@@ -0,0 +1,918 @@
...
+        break;
+      case POLLING_QMN:
+        this->state_ = STATE_IDLE;
+        break;
+    }
+  }
+
+  if (this->state_ == STATE_POLL_CHECKED) {
+    bool enabled = true;
+    String fc;
```
Comment (OttoWinter) [inline esphome/components/pipsolar/pipsolar.cpp:424]:
```suggestion
    std::string fc;
```

In esphome, we use the C++ stdlib `std::string` for strings. Please update the code to use that here (see here for api reference: https://en.cppreference.com/w/cpp/string/basic_string)

File: esphome/components/pipsolar/pipsolar.cpp [hunk 2]
```diff
@@ -0,0 +1,918 @@
...
+    char tmp[PIPSOLAR_READ_BUFFER_LENGTH];
+    sprintf(tmp, "%s", this->read_buffer_);
```
Comment (OttoWinter) [inline esphome/components/pipsolar/pipsolar.cpp:426]: Isn't this just copying the read buffer to the tmp variable? Why not just `sscanf()` off of the `read_buffer_` variable directly?

**Review by OttoWinter — COMMENTED** [review]
Thanks!

So this PR is pretty big ^^ Which seems to stem mostly from boilerplate code. If you can, please look into replacing some of the code on the C++ side with macros (see for example how `#define TEMPLATABLE` is defined). On the python side you could use `getattr()` like in `ade7953/sensor.py` to reduce some of the boilerplate code. It's ok too like this, but would be better with shorter files.

After a quick look the code looks ok - just one thing about using `std::string` instead of `String`

File: esphome/components/pipsolar/pipsolar.cpp (hunk 2 above)
Comment (andreashergert1984) [reply esphome/components/pipsolar/pipsolar.cpp:426]: sscanf does not take it: 
error: invalid conversion from 'uint8_t* {aka unsigned char*}' to 'const char*' [-fpermissive]
         sscanf(this->read_buffer_,

**Comment by andreashergert1984** [comment]
so String replaced with std::string

File: esphome/components/pipsolar/pipsolar.cpp (hunk 2 above)
Comment (andreashergert1984) [reply esphome/components/pipsolar/pipsolar.cpp:428]: > Thanks!
> 
> So this PR is pretty big ^^ Which seems to stem mostly from boilerplate code. If you can, please look into replacing some of the code on the C++ side with macros (see for example how `#define TEMPLATABLE` is defined). On the python side you could use `getattr()` like in `ade7953/sensor.py` to reduce some of the boilerplate code. It's ok too like this, but would be better with shorter files.

this is by a timely manner not possible at the moment for me.. i had to get into it ;)

**Comment by andreashergert1984** [comment]
> Thanks!
> 
> So this PR is pretty big ^^ Which seems to stem mostly from boilerplate code. If you can, please look into replacing some of the code on the C++ side with macros (see for example how `#define TEMPLATABLE` is defined). On the python side you could use `getattr()` like in `ade7953/sensor.py` to reduce some of the boilerplate code. It's ok too like this, but would be better with shorter files.
> 
needs to be done in a later release.. no time..

**Comment by cristi2005** [comment]
Hi, Any news about implement this adon in Esphome ?

**Comment by andreashergert1984** [comment]
> Hi, Any news about implement this adon in Esphome ?

Feel free to review it or push it on discord ;)

**Comment by ans-gw** [comment]
Very interesting,great Idea add on....waiting for EspHome implementation.
Thanks

**Comment by andreashergert1984** [comment]
> Very interesting,great Idea add on....waiting for EspHome implementation.
> Thanks

It is already implemented.. need to be QA and merged

thx, for the flowers

**Review by jesserockz — APPROVED** [review]

File: esphome/components/pipsolar/__init__.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+    yield uart.register_uart_device(var, config)
+
+
+@automation.register_action(
+    "Pipsolaroutput.set_level",
```
Comment (jesserockz) [inline esphome/components/pipsolar/__init__.py]:
```suggestion
    "pipsolaroutput.set_level",
```

File: esphome/components/pipsolar/__init__.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    SetOutputAction,
+    cv.Schema(
+        {
+            cv.Required(CONF_ID): cv.use_id(CONF_ID),
+            cv.Required(CONF_VALUE): cv.templatable(cv.positive_float),
+        }
+    ),
+)
+def Pipsolaroutput_set_level_to_code(config, action_id, template_arg, args):
```
Comment (jesserockz) [inline esphome/components/pipsolar/__init__.py]:
```suggestion
def pipsolaroutput_set_level_to_code(config, action_id, template_arg, args):
```

File: esphome/components/pipsolar/__init__.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    "pipsolaroutput.set_level",
+    SetOutputAction,
+    cv.Schema(
+        {
+            cv.Required(CONF_ID): cv.use_id(CONF_ID),
+            cv.Required(CONF_VALUE): cv.templatable(cv.positive_float),
+        }
+    ),
+)
+def pipsolaroutput_set_level_to_code(config, action_id, template_arg, args):
```
Comment (jesserockz) [inline esphome/components/pipsolar/__init__.py]:
```suggestion
    "output.pipsolar.set_level",
    SetOutputAction,
    cv.Schema(
        {
            cv.Required(CONF_ID): cv.use_id(CONF_ID),
            cv.Required(CONF_VALUE): cv.templatable(cv.positive_float),
        }
    ),
)
def output_pipsolar_set_level_to_code(config, action_id, template_arg, args):
```

**Comment by jesserockz** [comment]
I pushed some commits that significantly scale down the code size but should retain the exact same functionality. 
I cannot test this but it does compile given the example YAML in the PR description. Let me know if it doesn't work or you want to know more.

**Comment by andreashergert1984** [comment]
> I pushed some commits that significantly scale down the code size but should retain the exact same functionality.
> I cannot test this but it does compile given the example YAML in the PR description. Let me know if it doesn't work or you want to know more.

give me some days to test the changed code in real live ;)

File: esphome/components/pipsolar/pipsolar.cpp [hunk 6]
```diff
@@ -0,0 +1,922 @@
...
+      command_queue_[testposition] = command;
+      ESP_LOGD(TAG, "Command queued successfully: %s with length %lu at position %d", command,
+               command_queue_[testposition].length(), testposition);
+      return;
+    }
+  }
+  ESP_LOGD(TAG, "Command queue full dropping command: %s", command);
+}
+
+void Pipsolar::switch_command(std::string command) {
```
Comment (jesserockz) [inline esphome/components/pipsolar/pipsolar.cpp]:
```suggestion
void Pipsolar::switch_command(const std::string &command) {
```

File: esphome/components/pipsolar/pipsolar.h [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  PIPSOLAR_TEXT_SENSOR(last_qmn, QMN)
+
+  PIPSOLAR_SWITCH(output_source_priority_utility_switch, QPIRI)
+  PIPSOLAR_SWITCH(output_source_priority_solar_switch, QPIRI)
+  PIPSOLAR_SWITCH(output_source_priority_battery_switch, QPIRI)
+  PIPSOLAR_SWITCH(input_voltage_range_switch, QPIRI)
+  PIPSOLAR_SWITCH(pv_ok_condition_for_parallel_switch, QPIRI)
+  PIPSOLAR_SWITCH(pv_power_balance_switch, QPIRI)
+
+  void switch_command(std::string command);
```
Comment (jesserockz) [inline esphome/components/pipsolar/pipsolar.h]:
```suggestion
  void switch_command(const std::string &command);
```

**Comment by jesserockz** [comment]
@andreashergert1984 Did you give the latest a test?

**Comment by andreashergert1984** [comment]
> @andreashergert1984 Did you give the latest a test?

ok. tested in my env. still works. 

thx
and sry for the delay.. a lot to do.

Andi

**Comment by lucasimons** [comment]
I try to use this with my inverter PIP2424mse but it doesn't work anyone tested this ?

**Comment by andreashergert1984** [comment]
> I try to use this with my inverter PIP2424mse but it doesn't work anyone tested this ?

did you tried the QPIGS Textsensor? 
do you have a protocol description for you specific device?

Andi

**Comment by lucasimons** [comment]
I think that it is the same protocol...  but I don't receive nothing from the inverter, onother people test it with other inverter and there is a swapped value in qpigs https://github.com/syssi/esphome-pipsolar/issues/3

I have two wemos d1 mini to use for test one with a big rs232 converter and the other with a little converter

Which inverters has it been tested with?

**Comment by andreashergert1984** [comment]
> I think that it is the same protocol...  but I don't receive nothing from the inverter, onother people test it with other inverter and there is a swapped value in qpigs [syssi/esphome-pipsolar#3](https://github.com/syssi/esphome-pipsolar/issues/3)
> 
> I have two wemos d1 mini to use for test one with a big rs232 converter and the other with a little converter
> 
> Which inverters has it been tested with?

i use the max3232 chip to do the job. 
But also directly connection a USB to RS232 works.. you just have to "simulate" the checksum at the end of requests. 

potentially it is better to move this discussion to the discord as it is better for support. 
https://discord.gg/tNsqNxvz

**Comment by lucasimons** [comment]
I did not understand you connected via usb to the inverter and then you have an rs232 converter?  Anyway on discord should I open a new thread in general-support?  Can you open it and I'll add myself to it?

**Comment by andreashergert1984** [comment]
> I did not understand you connected via usb to the inverter and then you have an rs232 converter? Anyway on discord should I open a new thread in general-support? Can you open it and I'll add myself to it?
No i tested both.. using a usb to rs232 (with a pc) and the max3232 converter with the esp. 

please open a new thread an write also to me (zoodirektor). 
 but i will only have time later that day (germany)

**Pull request merged**